### DISCUSSION TITLE
How to find out the owners of Regions?

### DISCUSSION THREAD

**Discussion by Xazax-hun** [GabrielDosReis/ipr#254] [Q&A]
In https://github.com/GabrielDosReis/ipr/commit/59afd4298367bcd8a5ed0fd4efe1257ceb4a5f4d, `Region::owner()` was removed. We used this to check the context of the declarations. E.g., to check if a `Var` is a global, we traversed the parents  of the enclosing `Region`s and checked what kind of regions do we encounter. E.g., if we encounter a the `Region` of a `Block`, we know this is not a global variable. With  `Region::owner()` removed, we can no longer check what language construct do the parent regions of a `Var` correspond to. 

Is there an alternative way to get this information that I'm missing? Or are we supposed to maintain a mapping independently of IPR on the side?

**Comment by Xazax-hun** [comment]
Here is an example what was possible before. How to get the same functionality after removing `Region::owner()`?

```
    bool in_local_region_helper(const Region& region)
    {
        if (region.global())
            return false;

        switch (region.owner().category)
        {
        case Category_code::Block:
        case Category_code::For_in:
        case Category_code::Handler: return true;
        default: break;
        }

        return in_local_region_helper(region.enclosing());
    }

    bool is_local(const Var& var)
    {
        return ::in_local_region_helper(var.home_region());
    }
```


@GabrielDosReis

**Reply by GorNishanov** [reply]
Another case:

```c++
    IPR_MATCHER(is_method, Decl)
    {
        if (auto* func = util::view<Fundecl>(node))
        {
            if (util::view<Class>(func->home_region().owner()))
                return true;
        }
        return false;
    }
```
This particular case can be addressed separately from owner if we decide on whether IPR needs a method-type

**Reply by GorNishanov** [reply]
Another use case:

```c++
        const Udt* get_enclosing_udt_if_any(ParseContext& ctx)
        {
            auto* region = &ctx.current_region();
            for (;;)
            {
                switch (region->owner_sort())
                {
                case Node::Sort::Class:
                    return &cast<Class>(region->owner());
                case Node::Sort::Union:
                    return &cast<Union>(region->owner());

                case Node::Sort::Namespace:
                case Node::Sort::Block:
                    return nullptr;

                default:
                    region = &region->enclosing();
                }
            }
        }
```
